WhyApp/Extensions/CoreDataModelExtensions.swift

```diff
@@ -0,0 +1,101 @@
+//
+//  CoreDataModelExtensions.swift
+//  WhyApp
+//
+//  Created by Joriah Lasater on 7/4/18.
+//  Copyright © 2018 Joriah Lasater. All rights reserved.
+//
+
+import AVKit
+import UIKit
+
+extension Goal : XCoreDataProtocol { }
+extension Note : XCoreDataProtocol { }
+extension Image : XCoreDataProtocol { }
+extension Video : XCoreDataProtocol { }
+
+extension Goal {
+    func relationalObjects<T: XCoreDataProtocol>(type: T) -> [T] {
+        var result = [T]()
+        let set: NSSet?
+        
+        if type is Note {
+            set = self.notes
+        } else if type is Video {
+            set = self.videos
+        } else if type is Image {
+            set = self.images
+        } else {
+            set = NSSet()
+        }
+        
+        set?.forEach({ (element) in
+            if let element = element as? T {
+                result.append(element)
+            }
+        })
+        return result
+    }
+}
+
+extension Image {
+    func getUIImage() -> UIImage {
+        
+        guard let imageUrlPath = data else {
+            return UIImage()
+        }
+        
+        let nsDocumentDirectory = FileManager.default.temporaryDirectory
+        let imageURL = nsDocumentDirectory.appendingPathComponent(imageUrlPath)
+        
+        if let image = UIImage(contentsOfFile: imageURL.path) {
+            return image
+        }
+        
+        return UIImage()
+    }
+}
+
+extension Video {
+    
+    private func url() -> URL? {
+        guard let videoUrlPath = data else {
+            return nil
+        }
+        
+        let nsDocumentDirectory = FileManager.default.temporaryDirectory
+        
+        let videoUrl = nsDocumentDirectory.appendingPathComponent(videoUrlPath)
+        print("Video URL From CoreData")
+        print(videoUrl)
+        print("-----------------------------------")
+        return videoUrl
+    }
+    
+    func avPlayer() -> AVPlayer {
+        if let url = url() {
+            return AVPlayer(url: url)
+        }
+        return AVPlayer()
```

Comment:
I would reverse the control flow in this function by using a `guard` and returning the default `AVPlayer` if the url returns nil.

```swift
func avPlayer() -> AVPlayer {
    guard let url = url() else { return AVPlayer() }
    return AVPlayer(url: url)
}
```

If you wanted to be really clever, you could do this...

```swift
func avPlayer() -> AVPlayer {
    return url().flatMap(AVPlayer.init) ?? AVPlayer()
}
```

And I can explain that if you'd like.

Reply:
Now that is some syntax sugar! I know what the ?? does but I can't seem to recall the name